WifiART/src/userinterface/ZModel.java

```diff
@@ -62,9 +116,110 @@ public void setCallback(ZController fc) {
     }
 
 
-    // TODO(dmevans) Implement the encoding in toHexBuffer().
+
     @Override
     public String toHexBuffer() {
-        return null;
+        byte fieldsLength = FLOOR_INFO_LENGTH + HEIGHT_ABOVE_FLOOR_LENGTH + HEIGHT_ABOVE_FLOOR_UNCERTAINTY_LENGTH;
+        byte[] buffer = new byte[SUBELEMENT_ID_LENGTH + LENGTH_FIELD_LENGTH + fieldsLength];
+        buffer[SUBELEMENT_ID_INDEX] = SUBELEMENT_ID;
+        buffer[LENGTH_INDEX] = fieldsLength;
+        int floorInfo = getFloorInfoField(); // throws NullPointerException
+        fillLittleEndian(buffer, floorInfo, FLOOR_INFO_INDEX, FLOOR_INFO_LENGTH);
+        int heightAboveFloor = getHeightAboveFloorField();
+        fillLittleEndian(buffer, heightAboveFloor, HEIGHT_ABOVE_FLOOR_INDEX, HEIGHT_ABOVE_FLOOR_LENGTH);
+        int heightAboveFloorUncertainty = getHeightAboveFloorUncertaintyField();
+        fillLittleEndian(buffer, heightAboveFloorUncertainty, HEIGHT_ABOVE_FLOOR_UNCERTAINTY_INDEX, HEIGHT_ABOVE_FLOOR_UNCERTAINTY_LENGTH);
+
+
+        String result = "";
+        for (byte b : buffer) {
+            result += String.format("%02x", b); // Convert the byte to a hex string of 2 characters.
+        }
+        return result;
+    }
+
+    private int getFloorInfoField() throws NullPointerException {
+        int result;
+
+        String locationMovementString = state.getLocationMovement();
+        try {
+            result = ExpectedToMove.FIELD_VALUES_MAP.get(locationMovementString); // Bits 0 and 1
+        } catch (NullPointerException exception) {
```

Comment:
Is it possible to avoid NullPointerException here? The state should probably validate the location movement when setting it (or using enum).

Reply:
I do validate field values in the state when setting them, specifically when the user can enter bad inputs. In this case, the issue is when a value was never set to begin with, so calculating a buffer is impossible since none of the three JRadioButton options have been chosen. Validating when setting the value isn't necessary for a JRadioButton input since having JRadioButtons constrains possible inputs to those that are allowed.

Reply:
Just a thought - since the spec already defines an "unknown" state, does it make sense to treat "not choosing == unknown"? Or do we want user to choose this option explicitly?

Reply:
One approach is to define a sensible default (and state what default is used in the UI description, if not set).